File: generated-code/codecraft/kotlin/src/main/kotlin/model/PlayerView.kt
package model

import util.StreamUtil

/**
 * Information available to the player
 */
class PlayerView {
    /**
     * Your player's ID
     */
    var myId: Int = 0
    /**
     * Size of the map
     */
    var mapSize: Int = 0
    /**
     * Whether fog of war is enabled
     */
    var fogOfWar: Boolean = false
    /**
     * Entity properties for each entity type
     */
    lateinit var entityProperties: MutableMap<model.EntityType, model.EntityProperties>
    /**
     * Max tick count for the game
     */
    var maxTickCount: Int = 0
    /**
     * Max pathfind nodes when performing pathfinding in the game simulator
     */
    var maxPathfindNodes: Int = 0
    /**
     * Current tick
     */
    var currentTick: Int = 0
    /**
     * List of players
     */
    lateinit var players: Array<model.Player>
    /**
     * List of entities
     */
    lateinit var entities: Array<model.Entity>

    constructor(myId: Int, mapSize: Int, fogOfWar: Boolean, entityProperties: MutableMap<model.EntityType, model.EntityProperties>, maxTickCount: Int, maxPathfindNodes: Int, currentTick: Int, players: Array<model.Player>, entities: Array<model.Entity>) {
        this.myId = myId
        this.mapSize = mapSize
        this.fogOfWar = fogOfWar
        this.entityProperties = entityProperties
        this.maxTickCount = maxTickCount
        this.maxPathfindNodes = maxPathfindNodes
        this.currentTick = currentTick
        this.players = players
        this.entities = entities
    }

    /**
     * Write PlayerView to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, myId)
        StreamUtil.writeInt(stream, mapSize)
        StreamUtil.writeBoolean(stream, fogOfWar)
        StreamUtil.writeInt(stream, entityProperties.size)
        for (entityPropertiesEntry in entityProperties) {
            val entityPropertiesKey = entityPropertiesEntry.key
            StreamUtil.writeInt(stream, entityPropertiesKey.tag)
            val entityPropertiesValue = entityPropertiesEntry.value
            entityPropertiesValue.writeTo(stream)
        }
        StreamUtil.writeInt(stream, maxTickCount)
        StreamUtil.writeInt(stream, maxPathfindNodes)
        StreamUtil.writeInt(stream, currentTick)
        StreamUtil.writeInt(stream, players.size)
        for (playersElement in players) {
            playersElement.writeTo(stream)
        }
        StreamUtil.writeInt(stream, entities.size)
        for (entitiesElement in entities) {
            entitiesElement.writeTo(stream)
        }
    }

    /**
     * Get string representation of PlayerView
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("PlayerView { ")
        stringBuilder.append("myId: ")
        stringBuilder.append(myId)
        stringBuilder.append(", ")
        stringBuilder.append("mapSize: ")
        stringBuilder.append(mapSize)
        stringBuilder.append(", ")
        stringBuilder.append("fogOfWar: ")
        stringBuilder.append(fogOfWar)
        stringBuilder.append(", ")
        stringBuilder.append("entityProperties: ")
        stringBuilder.append(entityProperties)
        stringBuilder.append(", ")
        stringBuilder.append("maxTickCount: ")
        stringBuilder.append(maxTickCount)
        stringBuilder.append(", ")
        stringBuilder.append("maxPathfindNodes: ")
        stringBuilder.append(maxPathfindNodes)
        stringBuilder.append(", ")
        stringBuilder.append("currentTick: ")
        stringBuilder.append(currentTick)
        stringBuilder.append(", ")
        stringBuilder.append("players: ")
        stringBuilder.append("[ ")
        var playersIndex = 0
        for (playersElement in players) {
            if (playersIndex != 0) {
                stringBuilder.append(", ")
            }
            stringBuilder.append(playersElement)
            playersIndex++
        }
        stringBuilder.append(" ]")
        stringBuilder.append(", ")
        stringBuilder.append("entities: ")
        stringBuilder.append("[ ")
        var entitiesIndex = 0
        for (entitiesElement in entities) {
            if (entitiesIndex != 0) {
                stringBuilder.append(", ")
            }
            stringBuilder.append(entitiesElement)
            entitiesIndex++
        }
        stringBuilder.append(" ]")
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read PlayerView from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): PlayerView {
            var myId: Int
            myId = StreamUtil.readInt(stream)
            var mapSize: Int
            mapSize = StreamUtil.readInt(stream)
            var fogOfWar: Boolean
            fogOfWar = StreamUtil.readBoolean(stream)
            var entityProperties: MutableMap<model.EntityType, model.EntityProperties>
            val entityPropertiesSize = StreamUtil.readInt(stream)
            entityProperties = mutableMapOf();
            for (entityPropertiesIndex in 0 until entityPropertiesSize) {
                var entityPropertiesKey: model.EntityType
                entityPropertiesKey = model.EntityType.readFrom(stream)
                var entityPropertiesValue: model.EntityProperties
                entityPropertiesValue = model.EntityProperties.readFrom(stream)
                entityProperties.put(entityPropertiesKey, entityPropertiesValue)
            }
            var maxTickCount: Int
            maxTickCount = StreamUtil.readInt(stream)
            var maxPathfindNodes: Int
            maxPathfindNodes = StreamUtil.readInt(stream)
            var currentTick: Int
            currentTick = StreamUtil.readInt(stream)
            var players: Array<model.Player>
            players = Array(StreamUtil.readInt(stream), {
                var playersElement: model.Player
                playersElement = model.Player.readFrom(stream)
                playersElement
            })
            var entities: Array<model.Entity>
            entities = Array(StreamUtil.readInt(stream), {
                var entitiesElement: model.Entity
                entitiesElement = model.Entity.readFrom(stream)
                entitiesElement
            })
            return PlayerView(myId, mapSize, fogOfWar, entityProperties, maxTickCount, maxPathfindNodes, currentTick, players, entities)
        }
    }
}

File: generated-code/codecraft/cpp/model/PlayerView.cpp
#include "PlayerView.hpp"

PlayerView::PlayerView(int myId, int mapSize, bool fogOfWar, std::unordered_map<EntityType, EntityProperties> entityProperties, int maxTickCount, int maxPathfindNodes, int currentTick, std::vector<Player> players, std::vector<Entity> entities) : myId(myId), mapSize(mapSize), fogOfWar(fogOfWar), entityProperties(entityProperties), maxTickCount(maxTickCount), maxPathfindNodes(maxPathfindNodes), currentTick(currentTick), players(players), entities(entities) { }

// Read PlayerView from input stream
PlayerView PlayerView::readFrom(InputStream& stream) {
    int myId = stream.readInt();
    int mapSize = stream.readInt();
    bool fogOfWar = stream.readBool();
    size_t entityPropertiesSize = stream.readInt();
    std::unordered_map<EntityType, EntityProperties> entityProperties = std::unordered_map<EntityType, EntityProperties>();
    entityProperties.reserve(entityPropertiesSize);
    for (size_t entityPropertiesIndex = 0; entityPropertiesIndex < entityPropertiesSize; entityPropertiesIndex++) {
        EntityType entityPropertiesKey = readEntityType(stream);
        EntityProperties entityPropertiesValue = EntityProperties::readFrom(stream);
        entityProperties.emplace(std::make_pair(entityPropertiesKey, entityPropertiesValue));
    }
    int maxTickCount = stream.readInt();
    int maxPathfindNodes = stream.readInt();
    int currentTick = stream.readInt();
    std::vector<Player> players = std::vector<Player>();
    size_t playersSize = stream.readInt();
    players.reserve(playersSize);
    for (size_t playersIndex = 0; playersIndex < playersSize; playersIndex++) {
        Player playersElement = Player::readFrom(stream);
        players.emplace_back(playersElement);
    }
    std::vector<Entity> entities = std::vector<Entity>();
    size_t entitiesSize = stream.readInt();
    entities.reserve(entitiesSize);
    for (size_t entitiesIndex = 0; entitiesIndex < entitiesSize; entitiesIndex++) {
        Entity entitiesElement = Entity::readFrom(stream);
        entities.emplace_back(entitiesElement);
    }
    return PlayerView(myId, mapSize, fogOfWar, entityProperties, maxTickCount, maxPathfindNodes, currentTick, players, entities);
}

// Write PlayerView to output stream
void PlayerView::writeTo(OutputStream& stream) const {
    stream.write(myId);
    stream.write(mapSize);
    stream.write(fogOfWar);
    stream.write((int)(entityProperties.size()));
    for (const auto& entityPropertiesEntry : entityProperties) {
        const EntityType& entityPropertiesKey = entityPropertiesEntry.first;
        const EntityProperties& entityPropertiesValue = entityPropertiesEntry.second;
        stream.write((int)(entityPropertiesKey));
        entityPropertiesValue.writeTo(stream);
    }
    stream.write(maxTickCount);
    stream.write(maxPathfindNodes);
    stream.write(currentTick);
    stream.write((int)(players.size()));
    for (const Player& playersElement : players) {
        playersElement.writeTo(stream);
    }
    stream.write((int)(entities.size()));
    for (const Entity& entitiesElement : entities) {
        entitiesElement.writeTo(stream);
    }
}

// Get string representation of PlayerView
std::string PlayerView::toString() const {
    std::stringstream ss;
    ss << "PlayerView { ";
    ss << "myId: ";
    ss << myId;
    ss << ", ";
    ss << "mapSize: ";
    ss << mapSize;
    ss << ", ";
    ss << "fogOfWar: ";
    ss << fogOfWar;
    ss << ", ";
    ss << "entityProperties: ";
    ss << "{ ";
    size_t entityPropertiesIndex = 0;
    for (const auto& entityPropertiesEntry : entityProperties) {
        if (entityPropertiesIndex != 0) {
            ss << ", ";
        }
        const EntityType& entityPropertiesKey = entityPropertiesEntry.first;
        const EntityProperties& entityPropertiesValue = entityPropertiesEntry.second;
        ss << entityTypeToString(entityPropertiesKey);
        ss << ": ";
        ss << entityPropertiesValue.toString();
        entityPropertiesIndex++;
    }
    ss << " }";
    ss << ", ";
    ss << "maxTickCount: ";
    ss << maxTickCount;
    ss << ", ";
    ss << "maxPathfindNodes: ";
    ss << maxPathfindNodes;
    ss << ", ";
    ss << "currentTick: ";
    ss << currentTick;
    ss << ", ";
    ss << "players: ";
    ss << "[ ";
    for (size_t playersIndex = 0; playersIndex < players.size(); playersIndex++) {
        const Player& playersElement = players[playersIndex];
        if (playersIndex != 0) {
            ss << ", ";
        }
        ss << playersElement.toString();
    }
    ss << " ]";
    ss << ", ";
    ss << "entities: ";
    ss << "[ ";
    for (size_t entitiesIndex = 0; entitiesIndex < entities.size(); entitiesIndex++) {
        const Entity& entitiesElement = entities[entitiesIndex];
        if (entitiesIndex != 0) {
            ss << ", ";
        }
        ss << entitiesElement.toString();
    }
    ss << " ]";
    ss << " }";
    return ss.str();
}